### PR TITLE
Character data api

### PR DESCRIPTION
**Pull request by CurtisGreen**
## Character data API (or more lovingly known as cookie-io)
Added a file with a bunch of get/set type functions that serves as the API for the character data, rather than using the object directly
1. Allows for better error checking
1. Prevents strange issue where data in scenes that are persistent (audio) and data in scenes that open/close often (level/dashboard) are out of sync
1. Removed some unneeded data points from defaultData
1. Added a parent "level" object so it's easy to loop through level data
1. Put everything related to cookies/character data in the cookie-io.js file

### Fixes
1. Fixed an issue where varrock mine was overwriting varrock assets
1. Fixed an issue where quest completion finished early
1. Fixed an issue where buying an item in the shop that you already have throws an error and doesn't stack

### Other
1. Quest completion now triggers a song to play (random between two clips)
1. Initialized member variables at the top of the Shop class
1. Made the scene asset variable names consistent

### CODE REVIEW

File: src/cookie-io.js [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    }
+    setName(name) {
+        this.characterData.name = name;
+    }
+
+    addGold(amount) {
+        if (this.characterData.gold + amount >= 0) {
+            this.characterData.gold += amount;
+        } else {
+            console.log(
```
Comment (zengk95) [inline src/cookie-io.js:23]: console log bad

File: src/cookie-io.js [hunk 2]
```diff
@@ -0,0 +1,226 @@
+import { defaultData } from "./default-data.js";
+import { CONSTANTS } from "./constants/constants.js";
+
+function getDefaultData() {
+    // Reset data (deep copy)
+    return JSON.parse(JSON.stringify(defaultData));
+}
+
+class CharacterData {
```
Comment (zengk95) [inline src/cookie-io.js:9]: imagine using classes in 2020

File: src/cookie-io.js [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+                sum += this.characterData.levels[level].enemiesKilled[enemy];
+            }
+        }
+        return sum;
+    }
+
+    addTimesClicked(value) {
+        this.characterData.timesClicked += value;
+    }
+    getTimesClicked() {
```
Comment (zengk95) [inline src/cookie-io.js:57]: https://medium.com/@luke_smaki/javascript-es6-classes-8a34b0a6720a

you can do like 

get timesClicked() and then access it like character.timesClicked to make it more like other types of objects in javascript.

File: src/items/equipment.js [hunk 4]
```diff
@@ -103,7 +104,11 @@ export default class Equipment extends Item {
     // TODO: be able to have multiple different required levels for different skills
     checkRequiredLevel() {
         let skill = this.skill.toLowerCase();
-        let level = calcLevel(this.scene.characterData.skills[skill]);
-        return (level >= this.requiredLevel);
+        if (skill == "melee") {
```
Comment (zengk95) [inline src/items/equipment.js:107]: use ===

File: src/items/item.js [hunk 5]
```diff
@@ -96,17 +97,17 @@ export class Item extends ClickableObject {
     }
 
     async buy() {
```
Comment (zengk95) [inline src/items/item.js:99]: does this have to be async? not sure if you're using that correctly. most of the time it's used for IO like getting stuff from the disk or api calls

File: src/scenes/audio.js [hunk 6]
```diff
@@ -77,8 +74,9 @@ export class AudioScene extends Phaser.Scene {
 
     // Pause BGM while playing SFX
     playSfx(audioName) {
+        console.log("playing", audioName);
         this.currentSong.pause();
-        if (this.sfx != undefined && this.sfx.isPlaying) {
```
Comment (zengk95) [inline src/scenes/audio.js:81]: !==

File: src/utilities.js [hunk 7]
```diff
@@ -52,4 +30,4 @@ export async function getItemClass(itemName, scene) {
 
 export async function setItemClass(key, value) {
     itemClasses[key] = value;
-}
\ No newline at end of file
+}
```
Comment (zengk95) [inline src/utilities.js:33]: why is there so much code SMH

**Review by zengk95 — COMMENTED** [review]
utter spaghetti

File: src/cookie-io.js (hunk 2 above)
Comment (CurtisGreen) [reply src/cookie-io.js:9]: Shouldve used more spaghetti mb

File: src/cookie-io.js (hunk 1 above)
Comment (CurtisGreen) [reply src/cookie-io.js:23]: I'm going to make the chat window work as a functioning log later, so it's a good placeholder

File: src/cookie-io.js (hunk 3 above)
Comment (CurtisGreen) [reply src/cookie-io.js:57]: Didn't know they had static functions and stuff like that js, nice

File: src/utilities.js (hunk 7 above)
Comment (CurtisGreen) [reply src/utilities.js:33]: > why is there so much code SMH

It's a triple A title, what do you expect

File: src/items/item.js (hunk 5 above)
Comment (CurtisGreen) [reply src/items/item.js:99]: Need to be async to use await in line 104 `let boughtItem = await getItemClass(this.constructor.name, dashboard);`

**Pull request merged**